internal/route/api/v1/repo/path.go

```diff
@@ -0,0 +1,198 @@
+package repo
+
+import (
+	"encoding/base64"
+	"fmt"
+	"net/url"
+
+	"github.com/gogs/git-module"
+	"gogs.io/gogs/internal/context"
+	"gogs.io/gogs/internal/db/errors"
+)
+
+type GitRepoContent struct {
+	Type            string `json:"type"`
+	Target          string `json:"target,omitempty"`
+	Size            int64  `json:"size"`
+	Name            string `json:"name"`
+	Path            string `json:"path"`
+	Sha             string `json:"sha"`
+	URL             string `json:"url"`
+	GitURL          string `json:"git_url"`
+	HTMLURL         string `json:"html_url"`
+	DownloadURL     string `json:"download_url"`
+	Links           Links  `json:"_links"`
+	Content         string `json:"content,omitempty"`
+	SubmoduleGitURL string `json:"submodule_git_url,omitempty"`
+	Encoding        string `json:"encoding,omitempty"`
+}
+
+type Links struct {
+	Git  string `json:"git"`
+	Self string `json:"self"`
+	HTML string `json:"html"`
+}
+
+const (
+	// :base-url/:username/:project/raw/:refs/:path
+	templateDownloadURL = "%s/%s/%s/raw/%s"
+	// :base-url/repos/:username/:project/contents/:path
+	templateSelfLink = "%s/repos/%s/%s/contents/%s"
+	// :baseurl/repos/:username/:project/git/trees/:sha
+	templateGitURLLink = "%s/repos/%s/%s/trees/%s"
+	// :baseurl/repos/:username/:project/tree/:sha
+	templateHTMLLLink = "%s/repos/%s/%s/tree/%s"
+)
+
+func GetPathContents(c *context.APIContext) {
+	treeEntry, err := c.Repo.Commit.GetTreeEntryByPath(c.Repo.TreePath)
+	if err != nil {
+		c.NotFoundOrServerError("GetTreeEntryByPath", git.IsErrNotExist, err)
+	}
+
+	gitURL := fmt.Sprintf(templateGitURLLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), treeEntry.ID.String())
+	htmlURL := fmt.Sprintf(templateHTMLLLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), treeEntry.ID.String())
+	selfurl := fmt.Sprintf(templateSelfLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), c.Repo.TreePath)
+
+	gr := &GitRepoContent{
+		DownloadURL: fmt.Sprintf(templateDownloadURL, c.BaseURL, c.Params(":username"), c.Params(":reponame"), c.Repo.TreePath),
+		Size:        treeEntry.Size(),
+		Name:        treeEntry.Name(),
+		Path:        c.Repo.TreePath,
+		Sha:         treeEntry.ID.String(),
+		Links: Links{
+			Git:  gitURL,
+			Self: selfurl,
+			HTML: htmlURL,
+		},
+		URL:     selfurl,
+		GitURL:  gitURL,
+		HTMLURL: htmlURL,
+	}
+	// A tree entry can only be one of the following types:
+	// 1. Tree (Directory)
+	// 2. SubModule
+	// 3. SymLink
+	// 4. Blob
+	if treeEntry.IsDir() {
+		dirTree, err := c.Repo.GitRepo.GetTree(treeEntry.ID.String())
```

Comment:
I didn't check, does  `treeEntry` has `ListEntries` method? So we don't need to get the tree again.

Reply:
I just checked again, it  looks like it doesn't: Here https://github.com/gogs/git-module/blob/master/tree_entry.go or here: https://github.com/gogs/git-module/blob/f29bce1e3a666c02175d080892be185405dd3af4/tree_entry.go. 

It looks like only `Tree` has that method and that too only in this specific commit: https://github.com/gogs/git-module/blob/f29bce1e3a666c02175d080892be185405dd3af4/tree.go#L137